### PR TITLE
Open with right panel

### PR DESCRIPTION
**Pull request by will-moore**
# What this PR does

Adds support for Open-with to top of right panel.
See https://trello.com/c/Cdzu1KlJ/57-open-with-from-right-panel

This allows you to "Open-with" without using the tree right-click menu, particularly useful for Images in Wells, that you don't see in the tree.

# Testing this PR

1. In our current setup, we don't have any "open with" plugins that support any objects except Images.

2. Check that the Open_with button shows the list of open-with plugins (same as right-click on tree -> open with) and that they are enabled/disabled according to selected object(s) in the same way as in the tree.

3. Both disabled and enabled items should give an appropriate tooltip.

4. Clicking on enabled items should open link in new tab, or you can right-click on the item and choose browser option of "Open in new Window", as for any regular link.

5. Browse SPW -> Well -> Image. Check you can open image in iViewer or image(s) in figure.

6. Also, check that a new Open-with icon is used both in the tree menu and for the new button.

screen shot 2017-04-04 at 11 11 42

# Related reading

See https://trello.com/c/Cdzu1KlJ/57-open-with-from-right-panel

--depends-on #5285

### CODE REVIEW

**Comment by will-moore** [comment]
cc @waxenegger I noticed when using the config at https://github.com/ome/omero-iviewer/tree/master/plugin/omero_iviewer that this uses "images" as supported objects instead of a single "image" and it doesn't include the "label" for the UI.

I've opened a PR to update this on our management tools: https://github.com/openmicroscopy/management_tools/pull/358
```
$ bin/omero config append omero.web.open_with '["omero_iviewer", "omero_iviewer_index", {"supported_objects":["image"], "script_url": "omero_iviewer/openwith.js", "label": "OMERO.iviewer"}]'
```
and the same should be applied to https://github.com/ome/omero-iviewer/blob/master/plugin/omero_iviewer/README.rst
Actually, it would be nice to move this info to the top-level README (same as for https://github.com/ome/omero-figure) since it's tricky to find in /plugin/ and most users won't know to look there.

**Comment by jburel** [comment]
@will-moore The reason it is on the readme under ``plugin/omero_iviewer`` is b/c that's the readme used on pypi cf. https://testpypi.python.org/pypi?:action=display&name=omero-iviewer&version=0.0.7
i.e. there is another readme for ol3-viewer.

The organisation of the repo is different that the one of figure

**Comment by waxenegger** [comment]
Checked 1-6, didn't run into any problems.

**Comment by will-moore** [comment]
Robot test failing - needs updating...
https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/571/robot/Web%20&%20Web/Web_1/Center%20Right%20Panel%20Tests/Check%20Center%20and%20Right%20Panel%20Sync%20For%20Image/

**Comment by jburel** [comment]
Having the open with in right-hand panel make the "Full Viewer" button a bit redundant. Something to probably remove at some point.

**Comment by jburel** [comment]
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/574/ output from yesterday

**Comment by jburel** [comment]
The option should not be available for shares https://github.com/openmicroscopy/openmicroscopy/pull/5250
The option is not available when selecting the object in the tree

**Comment by will-moore** [comment]
Don't show the "Open with" button in right panel for images in share.
Robot failures are flakiness not related to this PR.

File: components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/annotations/includes/toolbar.html [hunk 1]
```diff
@@ -266,20 +329,33 @@
...
+              <span></span>
+          </button>
+          <ul id="right_panel_openwith" class="dropdown dropdown_right">
+          </ul>
+          {% endif %}
+      </div>
+  </div>
 
 </div>
 <div style="clear:both"></div>
```
Comment (jburel) [inline components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/annotations/includes/toolbar.html]: add extra line at the end

File: components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/annotations/includes/toolbar.html [hunk 2]
```diff
@@ -39,6 +39,69 @@
...
+                    if (typeof v.isEnabled === "function") {
+                        // If plugin has provided a function 'isEnabled'...
+                        // prepare json of selected objects to pass to function
+                        enabled = v.isEnabled(selJson);
+                    }
+                    // ...Otherwise if supported_objects list is configured...
+                    // v.supported_objects is ['image'] or ['dataset', 'images'] etc.
+                    else if (typeof v.supported_objects === "object" && v.supported_objects.length > 0) {
+                        enabled = v.supported_objects.reduce(function(prev, supported){
+                            // E.g. If supported_objects is 'images'...
```
Comment (jburel) [inline components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/annotations/includes/toolbar.html]: minor ``e.g.``

**Comment by jburel** [comment]
2 very minor comments.
The option is no longer available.

**Comment by jburel** [comment]
Something is not logic:
When a well is selected, the full viewer option is on and the image can be opened, but not the option in open_with

**Comment by will-moore** [comment]
When a Well is shown in the right panel, we've decided that we need to enable the Preview panel. Therefore it also made sense to enable the Viewer button (https://github.com/openmicroscopy/openmicroscopy/pull/5208#issuecomment-292189221). But, the open_with plugins only support "image" and are therefore disabled for "well".

The logic ```if 'well' then get the 'image'``` hasn't been applied to open_with plugins. What if a plugin supported 'well' and not 'image'? Or if it supported both, which would we use?

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#684](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/684/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/684/consoleText) for more details.
Possible conflicts:
  - PR #5208 will-moore 'Well preview panel'
    - components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/annotations/includes/toolbar.html

**Comment by jburel** [comment]
I understand the logic of the implementation but it think that from a user pov it is confusing.
let's discuss today
The easiest option will be to removed the "main viewer" from the list of plugins when used in the right-hand panel

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#685](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/685/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/685/consoleText) for more details.
Possible conflicts:
  - PR #5208 will-moore 'Well preview panel'
    - components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/annotations/includes/toolbar.html

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#686](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/686/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/686/consoleText) for more details.
Possible conflicts:
  - PR #5208 will-moore 'Well preview panel'
    - components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/annotations/includes/toolbar.html

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#687](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/687/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/687/consoleText) for more details.
Possible conflicts:
  - Upstream changes
    - components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/annotations/includes/toolbar.html

**Comment by jburel** [comment]
I can leave with the "divergence" for well since the actual viewer will be replaced by the new viewer
I will add a card to add support for well in iviewer
it will be good if a well is selected to have all the fields on the left hand side

**Comment by jburel** [comment]
card added https://trello.com/c/kNh9B4cX/4-open-with-more-target

**Pull request merged**